### PR TITLE
Escaping and sanitization.

### PR DESCRIPTION
**Pull request by andrewlimaza**
Escaping and sanitizing where possible.

### CODE REVIEW

File: includes/pmpro-gift-levels.php [hunk 1]
```diff
@@ -3,7 +3,7 @@
  * Load Email Template
  */
 function pmprowoo_gift_levels_email_path($default_templates, $page_name, $type = 'email', $where = 'local', $ext = 'html') {
-  $default_templates[] = PMPROWC_DIR . "/email/{$page_name}.{$ext}";
+  $default_templates[] = esc_url( PMPROWC_DIR . "/email/{$page_name}.{$ext}" );
```
Comment (andrewlimaza) [inline includes/pmpro-gift-levels.php:6]: This one needs checking, I think esc_url is right here, maybe esc_raw_url?

File: includes/pmpro-gift-levels.php [hunk 2]
```diff
@@ -150,10 +150,10 @@ function pmprowoo_gift_levels_render_on_cart_and_checkout( $cart_data, $cart_ite
         $custom_items = $cart_data;
     }
     if( isset( $cart_item['gift_recipient_name'] ) ) {
-        $custom_items[] = array( "name" => __( 'Recipient Name', 'pmpro-woocommerce' ), "value" => $cart_item['gift_recipient_name'] );
+        $custom_items[] = array( "name" => esc_html__( 'Recipient Name', 'pmpro-woocommerce' ), "value" => esc_html( $cart_item['gift_recipient_name'] ) );
```
Comment (ideadude) [inline includes/pmpro-gift-levels.php:153]:
```suggestion
        $custom_items[] = array( "name" => esc_html__( 'Recipient Name', 'pmpro-woocommerce' ), "value" => $cart_item['gift_recipient_name'] );

```

File: includes/pmpro-gift-levels.php [hunk 3]
```diff
@@ -150,10 +150,10 @@ function pmprowoo_gift_levels_render_on_cart_and_checkout( $cart_data, $cart_ite
         $custom_items = $cart_data;
     }
     if( isset( $cart_item['gift_recipient_name'] ) ) {
-        $custom_items[] = array( "name" => __( 'Recipient Name', 'pmpro-woocommerce' ), "value" => $cart_item['gift_recipient_name'] );
+        $custom_items[] = array( "name" => esc_html__( 'Recipient Name', 'pmpro-woocommerce' ), "value" => esc_html( $cart_item['gift_recipient_name'] ) );
     }
     if( isset( $cart_item['gift_recipient_email'] ) ) {
-        $custom_items[] = array( "name" => __( 'Recipient Email', 'pmpro-woocommerce' ), "value" => $cart_item['gift_recipient_email'] );
+        $custom_items[] = array( "name" => esc_html__( 'Recipient Email', 'pmpro-woocommerce' ), "value" => esc_html( $cart_item['gift_recipient_email'] ) );
```
Comment (ideadude) [inline includes/pmpro-gift-levels.php:156]:
```suggestion
        $custom_items[] = array( "name" => esc_html__( 'Recipient Email', 'pmpro-woocommerce' ), "value" => $cart_item['gift_recipient_email'] );

```

File: pmpro-woocommerce.php [hunk 4]
```diff
@@ -182,17 +182,17 @@ function pmprowoo_add_membership_from_order( $order_id ) {
 				
 				//create custom level to mimic PMPro checkout
 				$custom_level = array(
-					'user_id'         => $user_id,
-					'membership_id'   => $pmpro_level->id,
+					'user_id'         => (int) $user_id,
+					'membership_id'   => sanitize_text_field( $pmpro_level->id ),
```
Comment (ideadude) [inline pmpro-woocommerce.php:186]:
```suggestion
					'membership_id'   => $pmpro_level->id,
```

File: pmpro-woocommerce.php [hunk 5]
```diff
@@ -182,17 +182,17 @@ function pmprowoo_add_membership_from_order( $order_id ) {
 				
 				//create custom level to mimic PMPro checkout
 				$custom_level = array(
-					'user_id'         => $user_id,
-					'membership_id'   => $pmpro_level->id,
+					'user_id'         => (int) $user_id,
```
Comment (ideadude) [inline pmpro-woocommerce.php:185]:
```suggestion
					'user_id'         => $user_id,
```

File: pmpro-woocommerce.php [hunk 6]
```diff
@@ -182,17 +182,17 @@ function pmprowoo_add_membership_from_order( $order_id ) {
...
+					'membership_id'   => sanitize_text_field( $pmpro_level->id ),
 					'code_id'         => '', //will support PMPro discount codes later
-					'initial_payment' => $item['line_total'],
+					'initial_payment' => sanitize_text_field( $item['line_total'] ),
```
Comment (ideadude) [inline pmpro-woocommerce.php:188]:
```suggestion
					'initial_payment' => $item['line_total'],
```

File: pmpro-woocommerce.php [hunk 7]
```diff
@@ -182,17 +182,17 @@ function pmprowoo_add_membership_from_order( $order_id ) {
...
 					'billing_amount'  => '',
 					'cycle_number'    => '',
 					'cycle_period'    => '',
 					'billing_limit'   => '',
 					'trial_amount'    => '',
 					'trial_limit'     => '',
-					'startdate'       => $startdate,
+					'startdate'       => sanitize_text_field( $startdate ),
```
Comment (ideadude) [inline pmpro-woocommerce.php:195]:
```suggestion
					'startdate'       => $startdate,
```

File: pmpro-woocommerce.php [hunk 8]
```diff
@@ -564,23 +564,23 @@ function pmprowoo_tab_options() {
 			woocommerce_wp_checkbox(
 				array(
 					'id'          => '_membership_product_autocomplete',
-					'label'       => __( 'Autocomplete Order Status', 'pmpro-woocommerce' ),
-					'description' => __( "Check this to mark the order as completed immediately after checkout to activate the associated membership.", 'pmpro-woocommerce' ),
-					'cbvalue'	  => $cbvalue,
+					'label'       => esc_html__( 'Autocomplete Order Status', 'pmpro-woocommerce' ),
+					'description' => esc_html__( "Check this to mark the order as completed immediately after checkout to activate the associated membership.", 'pmpro-woocommerce' ),
+					'cbvalue'	  => esc_attr( $cbvalue ),
```
Comment (ideadude) [inline pmpro-woocommerce.php:569]:
```suggestion
					'cbvalue'	  => $cbvalue,
```

**Review by ideadude — COMMENTED** [review]
Removed some escaping when passing values to functions that escaped later when the function rendered output.

Let stuff like esc_html__ in since we don't trust localizers and effectively use esc_html to sanitize the text coming in.

File: includes/pmpro-gift-levels.php (hunk 1 above)
Comment (ideadude) [reply includes/pmpro-gift-levels.php:6]: sanitize_file_name
```suggestion
  $default_templates[] = PMPROWC_DIR . '/email/' . sanitize_file_name( $page_name . '.' . $ext );

```

**Pull request merged**